### ISSUE TITLE
HAL supports event reset on the device side which doesn't exist in CUDA

### ISSUE THREAD

**Issue by ThomasRaoux** [iree-org/iree#4869]
CUDA events can only be reset from the host using `cuEventRecord` (https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EVENT.html). 
HAL API exposes a way to reset an event on the device analogue to Vulkan [vkCmdResetEvent](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkEvent.html). 
The HAL interface has the following signature:
```
iree_status_t iree_hal_cuda_graph_command_buffer_reset_event(
    iree_hal_command_buffer_t* base_command_buffer, iree_hal_event_t* event,
    iree_hal_execution_stage_t source_stage_mask)
```

We won't be able to reasonably implement it for CUDA backend.
Are HAL events still being used right? What is the future plan for them? What would it take to remove this part of the API from HAL?

**Comment by ThomasRaoux** [comment]
FYI @benvanik

**Comment by benvanik** [comment]
TLDR: ignore reset, or make wait_events fully synchronize and ignore set/reset entirely as that's valid - just suboptimal for when we start using them. Future you (or someone) can figure out how to use them after things are brought up :)

Some background/musings:

HAL events are critical for communicating the dependency information through the HAL (it's how you define the edges on the flattened DAG) - but that doesn't mean it needs to map 1:1 with the underlying HAL implementation 
primitives (like CUevent). Think of the iree_hal_event_t as just a virtual handle to your actual CUDA event assigned in your implementation as needed. This means that iree_hal_event_t could just be an integer that maps into some pending CUDA event table and that is evaluated only during submission time (*not* during command buffer recording). Just a token to carry around so you know what you're referring to.

We need the events to have persistent handles as they can be used across command buffers (within the same queue aka CUDA stream) and those command buffers can be reused - if the events had to be different each time it would require re-recording. The indirection afforded by iree_hal_event_t lets us record command buffers once with a single event handle (essentially just an identifier) and replay them multiple times possibly with different backing event handles.

There is an important implicit guarantee in the event behavior we can exploit: since events are only usable within the same queue, and all submissions to a queue must be able to make forward progress (meaning that you can't issue a wait and _then_ a set to a queue - the sets must always come before waits), you then know that there will never be a reset->wait and instead always a reset->set->wait. The reset is important here because it is observable host behavior but not one we rely on today. We don't have a iree_hal_event_query that enables observation from the host yet - and maybe never as it has implications we don't want to expose to our compiled code (for example when running over RPC/sandbox/non-shared memory systems it is not cheap). So ignoring reset in the CUDA backend is totally valid. We can't remove the reset_event from the HAL though as it's required for vulkan interop - recording a set in vulkan does not reset the event and instead will no-op on an already set event leading to all kinds of badness. So we'll always emit the resets - even if they are ignored by particular backends.

CUDA graphs kind of mash all these concepts together: the events you assign are recorded before/after nodes execute and thus are implicitly reset and reused but it's not something you can do yourself from the API. I feel like that's a big compromise they made as it reduces CPU<->GPU sync opportunities: cuEventQuery and cuEventSynchronize now have some implicit requirements about when they are valid to call where you can't observe the pending state of an event until you have recorded it (submitted the graph) and need an external mechanism to order that host-side behavior. But again since we don't expose cuEventQuery/vkGetEventStatus or cuEventSynchronize (as there is no Vulkan equivalent) we're fine to ignore it :)

**Comment by ThomasRaoux** [comment]
Thanks a lot for the detailed explanation! If I understand well this means Events are used only for GPU<->GPU synchronizations within the same stream/queue. Does that mean I can just translate those into CUDA graph edges for the events within a command buffer? Within the same stream different command buffers are serialized so I assume ignoring events spanning across command buffer should be correct (even if at some pointer we might want to avoid the serialization). 

One implementation could be to keep track of all the leaf nodes when an event is signaled and use those as dependency nodes when waiting on the event, the reset would then just clear the book keeping. Would that work or am I missing something?
The part I'm still unclear about how we use Event vs Barriers to synchronize within a command buffer?

**Comment by benvanik** [comment]
Spot on, GPU<->GPU on the same device and same queue only. Semaphores are our coarser-grained primitive that allows CPU<->GPU and GPU<->GPU across devices or queues within a single device.

Good idea with the implementation; using a blackboard is what I do on the CPU implementation though I haven't gone too far on that yet and we could maybe share notes/utils whenever we get to doing it. I just accumulate dispatches until there's a barrier of any kind and flush them out - correct, but non-optimal. I know that I won't be writing the code to emit anything more than basic barriers from the compiler for awhile so it was a nice way to get some concurrency without needing to spend a lot of time up front on it :)

Barriers define a relationship between all previously recorded commands from all command buffers submitted to the same queue and all future recorded commands. Events can be thought of as "split barriers" - that's what they are referred to as in the Vulkan spec. They just give you a way to say "everything before this set_event has a _happens-before_ relationship with everything after this wait_events" whereas barriers are like a combined set/wait saying "everything before the barrier _happens-before_ everything after the barrier."

The way you fill bubbles in the pipeline is by packing as much work as you can between set_event/wait_events. That's what we'll be doing in the compiler when we use the SSA DAG to topologically sort dependencies ("move the users of %0 as far down as possible, set the event, move anything not transitively dependent on %0 to after the set_event, and then wait on the event before continuing with users of %0"). There's a tradeoff there with memory consumption: you need to keep the storage of %0 alive until it is fully consumed and something naive may end up going a bit too wide. This is an area where heuristics/profile-guided tuning/search can help.

There's also overhead to tracking the events and there may be actual execution overhead on the target device too so it's a tradeoff as to when to use the simpler barriers vs events. But you can imagine a canonicalizer on `hal.command_buffer.wait_events` that looks to see if the `hal.command_buffer.set_event` is right before it and if so turns it into a combined `hal.command_buffer.*_barrier` - we'll have that eventually but it may be a helpful way to reason about the differences today :)

**Comment by benvanik** [comment]
Just some FYI as this issue will be good to point people at in the future:

The other things barriers/events do is allow us to indicate desired execution and cache behavior. This allows for (but does not require) fine-grained barriers.

For example a barrier (or event set/wait pair) with a source scope of DISPATCH_WRITE and a target scope of INDIRECT_COMMAND_READ says "all previous writes from dispatch operations must be visible to all future indirect dispatches at the time their parameters are read" which just ensures that they happen in the correct order (execution dependency). memory_barriers are that on a global level (all device/host memory) while buffer_barriers let you scope it to particular buffers (only those you are reading back, etc).

Vulkan uses some specific terminology here; "to make available" effectively means "flush caches back to memory" and "to make visible" means "invalidate caches to cause fetches from memory." A barrier that has source scope of DISPATCH_WRITE and target of HOST_READ means that all memory written by dispatches prior to the barrier (or event set) must be made available to the host (flushed from the device) before any more commands can be executed but does not require it to be visible: the host may still have cached values and needs to invalidate its cache. The guarantee is only that if the host does invalidate its cache after the barrier it will see all of the memory written by dispatches on the device.

Another example is the reverse: HOST_WRITE to DISPATCH_READ would say that all host writes before the barrier must be made visible to all future dispatch read operations.

All this is another area, though, that allows for dramatic simplification: it's legal to completely ignore the memory_barriers/buffer_barriers if memory is always coherent, or always flush/invalidate as required all memory instead of scoping it. The only load-bearing part is queue family transfer but that's a whole can of worms I'd like to keep sealed for awhile :P (that's what will help dedicated DMA transfer queues run overlapped with execution queues at a fine granularity)

A good resource for examples, with this one like the above GPU->CPU readback: https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples#cpu-read-back-of-data-written-by-a-compute-shader

Note that what we are emitting from the compiler now is full pipeline barriers after each dispatch; I like the disclaimer here: https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples#full-pipeline-barrier :P

**Comment by benvanik** [comment]
I don't think there's anything to do here besides implement events in the CUDA HAL.
